### ISSUE TITLE
Can't process certain cookbooks

### ISSUE THREAD

**Issue by stuartphilp** [berkshelf/berkshelf-api#112]
Occasionally the cache builder will fail to process cookbooks with invalid characters in their metadata.json:

``` 
Processing metadata for 500 cookbooks with 5278 remaining on opscode
2014-04-24_18:38:09.38282 I, [2014-04-24T13:38:00.353518 #8728]  INFO -- : Saving the cache to: /etc/berkshelf/api-server/cerch
2014-04-24_18:38:09.38283 I, [2014-04-24T13:38:00.353977 #8728]  INFO -- : Cache saved!
2014-04-24_18:38:09.38284 E, [2014-04-24T13:38:09.381809 #8728] ERROR -- : Actor crashed!
2014-04-24_18:38:09.38285 Encoding::InvalidByteSequenceError: "\xE2" on US-ASCII
2014-04-24_18:38:09.38285       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/json-1.8.1/lib/json/common.rb:155:in `encode'
2014-04-24_18:38:09.38285       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/json-1.8.1/lib/json/common.rb:155:in `initialize'
2014-04-24_18:38:09.38285       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/json-1.8.1/lib/json/common.rb:155:in `new'
2014-04-24_18:38:09.38285       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/json-1.8.1/lib/json/common.rb:155:in `parse'
2014-04-24_18:38:09.38286       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/ridley-3.1.0/lib/ridley/chef/cookbook/metadata.rb:458:in `from_json'
2014-04-24_18:38:09.38286       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/ridley-3.1.0/lib/ridley/chef/cookbook/metadata.rb:29:in `from_json'
2014-04-24_18:38:09.38286       /opt/berkshelf-api/v1.3.1/lib/berkshelf/api/cache_builder/worker/opscode.rb:59:in `load_metadata'
2014-04-24_18:38:09.38286       /opt/berkshelf-api/v1.3.1/lib/berkshelf/api/cache_builder/worker/opscode.rb:39:in `block in metadata'
2014-04-24_18:38:09.38287       /opt/chef/embedded/lib/ruby/1.9.1/tmpdir.rb:83:in `mktmpdir'
2014-04-24_18:38:09.38287       /opt/berkshelf-api/v1.3.1/lib/berkshelf/api/cache_builder/worker/opscode.rb:37:in `metadata'
2014-04-24_18:38:09.38287       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/celluloid-0.16.0.pre/lib/celluloid/calls.rb:26:in `public_send'
2014-04-24_18:38:09.38287       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/celluloid-0.16.0.pre/lib/celluloid/calls.rb:26:in `dispatch'
2014-04-24_18:38:09.38288       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/celluloid-0.16.0.pre/lib/celluloid/calls.rb:63:in `dispatch'
2014-04-24_18:38:09.38288       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/celluloid-0.16.0.pre/lib/celluloid/cell.rb:60:in `block in invoke'
2014-04-24_18:38:09.38288       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/celluloid-0.16.0.pre/lib/celluloid/cell.rb:71:in `block in task'
2014-04-24_18:38:09.38288       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/celluloid-0.16.0.pre/lib/celluloid/actor.rb:362:in `block in task'
2014-04-24_18:38:09.38289       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/celluloid-0.16.0.pre/lib/celluloid/tasks.rb:55:in `block in initialize'
2014-04-24_18:38:09.38289       /opt/berkshelf-api/v1.3.1/vendor/bundle/ruby/1.9.1/gems/celluloid-0.16.0.pre/lib/celluloid/tasks/task_fiber.rb:15:in `block in create'
```

Perhaps the [opscode worker](https://github.com/berkshelf/berkshelf-api/blob/master/lib/berkshelf/api/cache_builder/worker/opscode.rb#L60) could handle this more gracefully

**Comment by miketheman** [comment]
Not sure if this is the right thread/repo to report this, but it looks like this is the same error message, so I'm continuing the report with details.

Using Ruby 2.2.0, berkshelf-api-client 1.2.1, berkshelf 3.2.3, ridley 4.1.2.

Confirming crash from [this cookbook's metadata](https://github.com/mthssdrbrg/kafka-cookbook/blob/82c3bdae2b19f31077cf0a1356eaa6ed2078ef78/metadata.rb#L4) as well as unicode in the [readme being slurped in here](https://github.com/jarosser06/chef-sysdig/blame/master/README.md#L69-L78).

Let's ignore the fact that unicode quotes are ugly, and that [should be a fix in](https://github.com/jarosser06/chef-sysdig/pull/3) and of itself. But allowing people to use their own name spelling seems like a reasonable allowance, right? :grin: 

Stack:
```console
Encoding::InvalidByteSequenceError: "\xE2" on US-ASCII
	$GEM_HOME/gems/json-1.8.2/lib/json/common.rb:155:in `encode'
	$GEM_HOME/gems/json-1.8.2/lib/json/common.rb:155:in `initialize'
	$GEM_HOME/gems/json-1.8.2/lib/json/common.rb:155:in `new'
	$GEM_HOME/gems/json-1.8.2/lib/json/common.rb:155:in `parse'
	$GEM_HOME/gems/ridley-4.1.2/lib/ridley/chef/cookbook/metadata.rb:473:in `from_json'
	$GEM_HOME/gems/ridley-4.1.2/lib/ridley/chef/cookbook/metadata.rb:29:in `from_json'
	$GEM_HOME/gems/ridley-4.1.2/lib/ridley/chef/cookbook.rb:36:in `from_path'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/cached_cookbook.rb:15:in `from_store_path'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/cookbook_store.rb:86:in `cookbook'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/cookbook_store.rb:67:in `import'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/cookbook_store.rb:30:in `import'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/installer.rb:106:in `block in install'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/downloader.rb:38:in `block in download'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/downloader.rb:35:in `each'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/downloader.rb:35:in `download'
	$GEM_HOME/gems/berkshelf-3.2.3/lib/berkshelf/installer.rb:105:in `install'
	$GEM_HOME/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in `public_send'
	$GEM_HOME/gems/celluloid-0.16.0/lib/celluloid/calls.rb:26:in `dispatch'
	$GEM_HOME/gems/celluloid-0.16.0/lib/celluloid/calls.rb:63:in `dispatch'
	$GEM_HOME/gems/celluloid-0.16.0/lib/celluloid/cell.rb:60:in `block in invoke'
	$GEM_HOME/gems/celluloid-0.16.0/lib/celluloid/cell.rb:71:in `block in task'
	$GEM_HOME/gems/celluloid-0.16.0/lib/celluloid/actor.rb:357:in `block in task'
	$GEM_HOME/gems/celluloid-0.16.0/lib/celluloid/tasks.rb:57:in `block in initialize'
	$GEM_HOME/gems/celluloid-0.16.0/lib/celluloid/tasks/task_fiber.rb:15:in `block in create'
```

I've tried mucking around with the local ENV, combination of these flags:
```
export LANG=en_US.UTF-8
export LANGUAGE=en_US.UTF-8
export LC_ALL=en_US.UTF-8
export RUBYOPTS="-E utf-8"
```

But no luck.

**Comment by miketheman** [comment]
(addendum) We get an older version of the berkshelf-api-client because anything newer is [blocked by Berkshelf](https://github.com/berkshelf/berkshelf/blob/9f311fb078fedae7d7a61a1f0b8eb86bb922f9e6/berkshelf.gemspec#L34).

**Comment by reset** [comment]
@miketheman what do you mean it's blocked by Berkshelf? The latest version of the API client is 1.2.1 and the constraint allows for `~> 1.2`

**Comment by miketheman** [comment]
> @miketheman what do you mean it's blocked by Berkshelf? The latest version of the API client is 1.2.1 and the constraint allows for ~> 1.2

My mistake - I was comparing this repo's releases instead of the api client repo - my bad. So please ignore the (addendum).

**Comment by reset** [comment]
@miketheman got ya. Yeah, the encoding stuff is still an issue that needs to be addressed.

**Comment by miketheman** [comment]
@reset I've tried the suggested workarounds that involve environment variables - nothing worked. Ideas?

**Comment by reset** [comment]
@miketheman I haven't taken the time to familiarize myself with the mess of text encoding and the Ruby VM. That's why encoding issues like this still exist. I'm not sure what the solution is because it seems to vary between versions of Ruby, different VMs, and how the data is being written or read.

**Comment by miketheman** [comment]
Sooo..... I spent a non-trivial amount of time figuring this out.
It's not berkshelf-api's fault, it's not berkshelf's fault, nor erubis, it's not even the JSON gem's fault.
It's not Ruby's fault either.

The solutions provided above, namely `export LANG=en_US.UTF-8` is likely to work in most cases.

HOWEVER, if in the event that the `en_US.UTF-8` locale **is not installed** no amount of exporting environment variables will get you there.

Turns out, that this is a [semi][1]-[known][2] [problem][3] with docker images, as they don't always contain much other than the base system and expect you to add whatever you like on top.

Once a locale exists, you can use `export` to your heart'c content, for my job I only needed `LANG` and no other environment variables to make this work.

The associated changes to the Dockerfile can be seen [here][4].

I thought about if there's a way to help surface this problem in any other manner, and working through the stacktrace it seems like since [ridley's parser for the metadata.rb][5] was the one that caught this.

The only thought I had was if ridley forced an `encoding: 'utf-8'` in the JSON.parse method, but since the json gem loads all strings [into UTF-8 by default][6], I don't think there's much to be gained there.

I don't think there's any real action to be taken here, other than shout this from the rooftops that if you run `locale -a` and don't see `en_US.UTF-8`, you might have a bad day.

[1]: http://jaredmarkell.com/docker-and-locales/
[2]: https://github.com/6si/docker-utf8
[3]: https://github.com/abevoelker/docker-ubuntu-locale
[4]: https://github.com/DataDog/docker-jenkins-slave/commit/ab85c65ed231322fe6a20d795ba493cf02cbafb6
[5]: https://github.com/reset/ridley/blob/598c81b2fb08a849d390738947eca8394db5e025/lib/ridley/chef/cookbook/metadata.rb#L473
[6]: https://github.com/flori/json/blame/5a4331918f61a8a4dbe12df8a12f66e5b4e4811a/README.rdoc#L16-L17

**Comment by ArturT** [comment]
Tip for MacOS:
```
export LANGUAGE=en_US.UTF-8
export LANG=en_US.UTF-8
export LC_ALL=en_US.UTF-8
```

**Comment by jkeiser** [comment]
The rest of Chef has moved on from the JSON gem, and I'm wondering if we should move this from JSON too.  @lamont-granquist, I dimly recall you were involved in that, can you shed some light / do you know who might be able to?

**Comment by reset** [comment]
@jkeiser @lamont-granquist which gem is Chef using now?

**Comment by miketheman** [comment]
@jkeiser As I think I figured out a bit above - it's not the JSON gem's faukt - it's the underlying system not having the locale installed.

**Comment by jkeiser** [comment]
@reset looks like YAJL FFI: https://github.com/chef/chef/blob/master/chef.gemspec#L25

@miketheman it could be that I've done my search wrong, but I'm not seeing this bug pop up with Chef, which also has to read cookbooks--this makes me wonder if the gem we switched to works around the issue in some way.

**Comment by jkeiser** [comment]
@reset the main reason for moving to the YAJL gem was that the error messages were a lot more user-friendly.  @daneilsdeleo was just regaling me with tales of that woe :)

**Comment by miketheman** [comment]
@jkeiser that's entirely possible. I don't recall digging deeper into the Chef codebase for this one - but it boiled down to when running a `JSON.parse` without having _any_ locales installed in a minimal docker image.
See here: https://github.com/DataDog/docker-jenkins-slave/commit/ab85c65ed231322fe6a20d795ba493cf02cbafb6

It's possible that the same behavior would be exhibited by Chef under the same circumstances - unless it's creating it's own `locale` settings.

**Comment by miketheman** [comment]
I suspect that @lamont-granquist is likely to have some opinions on how Chef interacts with locales - it appears that there's now some messaging to tell you to go and get one.
https://github.com/chef/chef/pull/2425/files#diff-ce8cfba0deb02929653a78f574664565R649

**Comment by jkeiser** [comment]
@miketheman awesome find.  I have a feeling Lamont is going to know whether we've worked around this in Chef or whether we're just as fucked as everybody else :)

**Comment by lamont-granquist** [comment]
lol

Yeah, so there's various ways to hose locales.  Easiest way is if someone just doesn't like the way `ls -la` sorts output in the `en_US.UTF_8` locale with dotfiles all scattered through the listing alphabetically -- and what people do is `export LC_ALL=C` or something like that and then they break UTF-8 for everything they run -- which works for 'muricans until they try to parse something with an umlaut in someone's name and then they think the software they ran is buggy (and that is exactly how I ran for years myself as a proud 'murican).  We also ran into people running on windows in Japan with some kind of strange non-UTF-8 charset which they had to have default on the system but which broke `knife upload` as well.

So, anyway, it turned out that we couldn't really fight the latter battle, and while we could fight the former battle that went nowhere because while locales aren't actually all that hard, its just hard enough that users don't care.  So what Chef 12 does now is force the default external encoding to UTF-8:

https://github.com/chef/chef/blob/68bab0b362c47e5a75492e57094f72ecee1171ae/lib/chef/application.rb#L193-L195

https://github.com/chef/chef/blob/master/lib/chef/config.rb#L698-L705

I still sort of think its philosophically wrong to be doing that, but in practice nobody seems to care and UTF-8 is mostly taking over the world, and by standing ground you just wind up in a mutual firing squad with users on one side and software devs on the other.   Plus if anyone ever does write a cookbook or whatever in Shift-JIS then they can still put a magic comment at the start of it to change the encoding.

If you look further into what Chef does with locales there's also this huge pile of code here which I do *not* think you need:

https://github.com/chef/chef/blob/master/lib/chef/config.rb#L646-L696

That has to do with the fact that we explicitly need an *English* locale because in many cases in providers we parse the output of commands and grep for localized strings and if the system is in `de_DE.UTF-8` you'll get back German which will break the providers -- so we need to force external commands to English and UTF-8 (except for when service providers are firing up someone's software which executes in the system locale).   I don't think you need that.

So, TL;DR add a line:

```
Encoding.default_external = Encoding::UTF_8
```

Make it configurable for bonus points, but chances are nobody will care.

**Comment by davetrend** [comment]
@jkeiser asked me to post this.  It started out as a chef-provisioning issue https://github.com/chef/chef-provisioning/issues/190, where Vagrant very early in its Berks vendoring throws the encoding exception out of the JSON gem -- all components up to date from chefdk 0.4.0, embedded ruby ruby 2.1.4p265,  vagrant 1.7.1.  @jkeiser tracked it down here; per our discussion, I looked at the locales of both my own machine, and the Centos 6.4 box I'm throwing up in Vagrant.  Interesting:

Here’s the host machine (mine, OSX 10.9)

PHUSCA-L24566:cookbooks treffda1$ env | grep LANG
    LANG=en_US.UTF-8
    PHUSCA-L24566:cookbooks treffda1$ locale
LANG="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_CTYPE="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_ALL=
PHUSCA-L24566:cookbooks treffda1$ locale -a | grep en_US.UTF-8
en_US.UTF-8
PHUSCA-L24566

Here’s a vagrant instance on that machine (our vanilla Centos 6.4 image)

[vagrant@localhost ~]$ env | grep LANG
LANG=en_US.UTF-8
[vagrant@localhost ~]$ locale
LANG=en_US.UTF-8
LC_CTYPE="en_US.UTF-8"
LC_NUMERIC="en_US.UTF-8"
LC_TIME="en_US.UTF-8"
LC_COLLATE="en_US.UTF-8"
LC_MONETARY="en_US.UTF-8"
LC_MESSAGES="en_US.UTF-8"
LC_PAPER="en_US.UTF-8"
LC_NAME="en_US.UTF-8"
LC_ADDRESS="en_US.UTF-8"
LC_TELEPHONE="en_US.UTF-8"
LC_MEASUREMENT="en_US.UTF-8"
LC_IDENTIFICATION="en_US.UTF-8"
LC_ALL=
[vagrant@localhost ~]$ locale -a | grep en_US.UTF-8
[vagrant@localhost ~]$ locale -a | grep en_US.utf8
en_US.utf8
[vagrant@localhost ~]$

Well well well.  Locale  en_US.UTF-8 isn’t installed; however, en_US.utf8 is.   Haven't yet tried setting the locale in /etc/sysconfig/i18n in the Vagrant instance and re-provisioning, that's next.

**Comment by hulu1522** [comment]
I am also getting this error using Chef 12.1.1 and Ruby 2.1.0 while running tests with docker.

I didn't want to clutter the comment but I posted my output here:
http://pastebin.com/W1sqXnK8

**Comment by klausmeyer** [comment]
> HOWEVER, if in the event that the en_US.UTF-8 locale is not installed no amount of exporting environment variables will get you there.
> 
> Turns out, that this is a semi-known problem with docker images, as they don't always contain much other than the base system and expect you to add whatever you like on top.

Had the same problem in a docker environment.

```shell
export LANG=C.UTF-8
```

Worked for me as a quick fix.

**Comment by lamont-granquist** [comment]
`C.UTF-8` isn't portable across O/Sen.  The only locale that I think is portable is the non-UTF8 'C' locale (kind of the root of this issue).

The ruby code in my TL;DR up there is portable though.

**Comment by klausmeyer** [comment]
That's why I mentioned it's a quick fix for me (and perhaps for others). Patching the ruby files wasn't an option in my case but I have to agree it is a nice(r) solution.

**Comment by guillaumemaka** [comment]
I had the same error, and fixed it with:
```
export LANG=en_US.UTF-8
export LC_CTYPE=en_US.UTF-8
```

```
$ chef -v
Chef Development Kit Version: 0.6.0
chef-client version: 12.3.0
berks version: 3.2.4
kitchen version: 1.4.0
```

**Comment by eric-tucker** [comment]
So this is still an issue and we encounter it when trying to fetch cookbooks using Berkshelf:

````
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.2/lib/json/common.rb:155:in `encode': "\xE2" on US-ASCII (Encoding::InvalidByteSequenceError)
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.2/lib/json/common.rb:155:in `initialize'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.2/lib/json/common.rb:155:in `new'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.2/lib/json/common.rb:155:in `parse'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/ridley-4.1.2/lib/ridley/chef/cookbook/metadata.rb:473:in `from_json'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/ridley-4.1.2/lib/ridley/chef/cookbook/metadata.rb:29:in `from_json'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/ridley-4.1.2/lib/ridley/chef/cookbook.rb:36:in `from_path'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cached_cookbook.rb:15:in `from_store_path'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cookbook_store.rb:108:in `block in cookbooks'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cookbook_store.rb:98:in `collect'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cookbook_store.rb:98:in `cookbooks'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/resolver/graph.rb:8:in `populate_store'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/resolver.rb:73:in `resolve'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/installer.rb:175:in `install_from_universe'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/installer.rb:39:in `run'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/berksfile.rb:374:in `install'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/berksfile.rb:583:in `block in vendor'
        from /opt/chefdk/embedded/lib/ruby/2.1.0/tmpdir.rb:88:in `mktmpdir'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/berksfile.rb:581:in `vendor'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cli.rb:387:in `vendor'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cli.rb:52:in `dispatch'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
        from /opt/chefdk/embedded/apps/berkshelf/lib/berkshelf/cli.rb:27:in `execute!'
        from /opt/chefdk/embedded/apps/berkshelf/bin/berks:5:in `<top (required)>'
        from /usr/bin/berks:43:in `load'
        from /usr/bin/berks:43:in `<main>'
````

Is this an issue with Berkshelf or the JSON gem?  Who do we look to in order to set a default LANG?

When I run through CodeDeploy, LANG is not set.  I'm not a fan of any application that cannot function (with reasonable defaults) without environment variables set as it breaks unattended execution.

**Comment by lamont-granquist** [comment]
@eric-tucker you need to learn how to set the locale of your system.  if you don't learn how to do that this is just the first of many issues you will have.  its like saying "i'm not a fan of any application that needs me to correctly setup DNS in /etc/resolv.conf, why can't it just work right?".  the locale is by definition external config that you have to apply because you're the one that knows the right locale and character set.  and what is right for you is wrong for someone in Europe using ISO-8851-1

**Comment by eric-tucker** [comment]
@lamont-granquist I agree.  I'm actually just asking for a reasonable default to allow unattended executions to function.  Of course there is a work-around so this isn't exactly a blocking issue.

I just think that with automation and unattended execution becoming more and more common place that relying on variables that are normally only set when a user is logged in is impeding progress from a DevOps standpoint.  Again, just **my opinion**.

**Comment by lamont-granquist** [comment]
From a systems engineering standpoint, you need to learn how to configure your locale.  Its your responsibility.  Every much your responsibility as configuring DNS, NTP, LDAP, your postgresql settings, etc.  And if CodeDeploy isn't setting LANG based on your system settings then it is the broken piece of software.  We're well into the 21st century now and to configure a server you need to care about localization, and if you can't learn how to do that you're simply not doing your job.

Just **my** opinion.

**Comment by eric-tucker** [comment]
@lamont-granquist I'm setting the locale in my CodeDeploy appspec.  This is how I can get the gem to work.

I feel like you're talking personal offence to this or arguing that I'm incompetent for feeling that environment variables should not be required for the gem to function.  Admittedly, I don't understand the deep logic of how the gem works, so a logical explanation why this application needs to have a value set for the locale is acceptable.

I admit I am biased against environment variables.  I'd rather see configuration placed in a file once and be done with it rather than having to set a bunch of environment variables every time I want to execute a program.  Of course there are exceptions.  However, to quote your example regarding `/etc/resolv.conf`, well that's a file.  Imagine if you had to set an environment variable to configure your DNS server for any application that had to do a DNS lookup.  That sounds crazy (at least to me).

If you feel that this is Amazon's fault for not having the LANG set during CodeDeploy (it actually is set when CloudFormation runs) then I will happily file a ticket with them.  The value is obviously getting set somewhere when CloudFormation runs and not because of a user logging into the system.  So this very well could be a deficiency in CodeDeploy.

**Comment by lamont-granquist** [comment]
Yes, the environment variables are absolutely necessary.  This is like how many apps fail when `$HOME` is not set correctly.  We now have to add the locale set of variables to that.  You can argue about that all you want but the design of that interface is done and its pretty futile to complain about environment variables being the interface.  They are the interface, and the design of that interface has nothing at all to do with berkshelf.  And yes, this is a bug that needs to get fixed in CodeDeploy and this is just the first of many issues that they'll have by breaking localization variables.

**Comment by lamont-granquist** [comment]
And yes I'm taking personal offense to your statement that "I'm not a fan of any application, [...etc...]" followed up by just blind bias and a demanding tone.  If you think it can be done better, than submit a PR request and show us how to do it.

**Comment by eric-tucker** [comment]
@lamont-granquist I was going to thank you for the thoughtful interface explanation as I finally see your point.  I was about to agree with you.  Then you had to follow-up with that last statement.

What did I do to provoke you?  Where have I been "demanding"?  Have I demanded that a default be set?  No.  I've asked if it would be possible and who we would look to in order to accomplish that if indeed possible.

You came in here stated that I had to learn how to do things, that I'm not doing my job, and giving me all sorts of sarcastic tone.  All because it is my opinion that things should work (minimally) without environment variables.  If my opinion is incorrect then factually demonstrate why rather than simply stating I need to learn to do things properly for the 21st century and the old line of "if you think you can do better do it yourself."

I'm not going to get into a flame war on GitHub on an issue that remains open for whatever reason.  If the solution is that things are working as intended and that a locale must be set via environment variables then the issue should be closed and discussion should stop and we move on.

**Comment by reset** [comment]
@eric-tucker a default being applied along with a warning being issued that the app is falling back to a default that may not be appropriate is what I think would close this issue

**Comment by signe** [comment]
On Yosemite, setting LC_ALL is the only variation that did the trick:

```
export LC_ALL=en_US.UTF-8
```

**Comment by camposer** [comment]
In my case changing LANG wasn't an option. Resolve my problem editing Berksfile as follows:

Encoding.default_external = "UTF-8"
source 'https://supermarket.chef.io'

metadata

After this, just executed berks install without any problem

**Comment by asottile** [comment]
I hit this in docker in an unrelated project and was able to solve it via:

```
RUN     /usr/sbin/locale-gen en_US.UTF-8
ENV     LANG en_US.UTF-8
```

**Comment by glennmate** [comment]
Identified and found this issue today. Usage of "export LC_CTYPE=en_US.UTF-8" resolved the issue. Prior to that, "env | Grep LC_CTYPE" was yielding no results. 

Thank you all.

**Comment by toddlers** [comment]
@guillaumemaka thanks :)

**Comment by dod38fr** [comment]
Weird, `berks verify` works with `LC_CTYPE=en_US.UTF-8` but fails with `LC_CTYPE=fr_FR.UTF-8`...

In more details:
```
$ export LC_CTYPE=en_US.UTF-8
$ berks verify
Verifying (40) cookbook(s)...
Verified.
$ export LC_CTYPE=fr_FR.UTF-8
-bash: warning: setlocale: LC_CTYPE: cannot change locale (fr_FR.UTF-8)
$ berks verify
/opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.3/lib/json/common.rb:155:in `encode': "\xC3" on US-ASCII (Encoding::InvalidByteSequenceError)
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.3/lib/json/common.rb:155:in `initialize'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.3/lib/json/common.rb:155:in `new'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/json-1.8.3/lib/json/common.rb:155:in `parse'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/ridley-4.5.1/lib/ridley/chef/cookbook/metadata.rb:570:in `from_json'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/ridley-4.5.1/lib/ridley/chef/cookbook/metadata.rb:29:in `from_json'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/ridley-4.5.1/lib/ridley/chef/cookbook.rb:36:in `from_path'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/cached_cookbook.rb:15:in `from_store_path'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/cookbook_store.rb:108:in `block in cookbooks'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/cookbook_store.rb:98:in `collect'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/cookbook_store.rb:98:in `cookbooks'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/cookbook_store.rb:163:in `satisfy'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/dependency.rb:139:in `cached_cookbook'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/lockfile.rb:116:in `block in trusted?'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/lockfile.rb:101:in `each'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/lockfile.rb:101:in `trusted?'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/berksfile.rb:737:in `validate_lockfile_trusted!'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/berksfile.rb:683:in `verify'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/cli.rb:404:in `verify'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/command.rb:27:in `run'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/invocation.rb:126:in `invoke_command'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor.rb:359:in `dispatch'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/cli.rb:52:in `dispatch'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/thor-0.19.1/lib/thor/base.rb:440:in `start'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/lib/berkshelf/cli.rb:27:in `execute!'
        from /opt/chefdk/embedded/lib/ruby/gems/2.1.0/gems/berkshelf-4.3.5/bin/berks:5:in `<top (required)>'
        from /usr/bin/berks:45:in `load'
        from /usr/bin/berks:45:in `<main>'
$ export LC_CTYPE=en_US.UTF-8
$ berks verify
Verifying (40) cookbook(s)...
Verified.
$ berks --version
4.3.5
```

Hope this helps

**Comment by dod38fr** [comment]
Forget my previous comment. There was an issue with my locale setup.

Sorry for the noide
